### DISCUSSION TITLE
Type instantiation is excessively deep and possibly infinite.

### DISCUSSION THREAD

**Discussion by RomanNabukhotnyi** [honojs/hono#2380] [Q&A] [answered]
Hey! I have a large application using hono and I'm running into a problem with types.

[Repo with a reproduction](https://github.com/cellajs/hono-case)

In the `main` branch I moved all my routes and added one additional test route and then the error appeared.

<img width="527" alt="image" src="https://github.com/honojs/hono/assets/97584054/6e377d31-c9c6-4308-85ed-16cfd33f5919">
<img width="752" alt="image" src="https://github.com/honojs/hono/assets/97584054/379eb41b-76bc-48cf-ad55-7602a5f7f831">

In the `test` branch I moved all the routes from the modules into one file and was able to add 35 additional routes without this error.

I also managed to compare the difference between Hono and Elysia
With Elysia I was able to manage 1410 routes (470 routes / 3 instances) and with Hono - 120 (40 / 3).
[Repo](https://github.com/cellajs/hono-elysia-test)
I might miss or forget something here, in that case please let me know.

**Comment by yusukebe** [comment]
Hi @RomanNabukhotnyi !

Thanks for using Hono.

This problem is caused by the large number of routes. Elysia supports many routes, I'll see if I can improve Hono as well.

**Reply by flipvh** [reply]
Thank you @yusukebe ! As I can imagine this isn't trivial, you have perhaps ideas or tips to temporarily bypass this limit? Can we limit the RPC functionality or do we have to remove it completely for now, so we can continue to add more routes? Also, let me know if you think we can help. I could post a cleaner reproduction if that can help you? Please let me know, thnks again

**Reply by yusukebe** [reply]
Hi @flipvh 

I'm struggling to improve the performance. Perhaps some improvement can be done by changing Types and fine tuning. But not easy to finish right now. So, please wait a moment.

As for bypassing the limit you mentioned, it would be a good idea to make the app smaller and create a client for each one.

For example, if you have the following application:

```ts
const appA = new Hono()
  .get('/1', (c) => c.json({ foo: 'bar ' }))
  .get('/2', (c) => c.json({ foo: 'bar ' }))
//...

const appB = new Hono()
  .get('/1', (c) => c.json({ foo: 'bar ' }))
  .get('/2', (c) => c.json({ foo: 'bar ' }))
//...

const oneApp = new Hono().route('/a', appA).route('/b', appB)
const client = hc<typeof oneApp>('http://localhost')
```

Just pass each app type to each client as follows:

```ts
const appA = new Hono()
  .get('/1', (c) => c.json({ foo: 'bar ' }))
  .get('/2', (c) => c.json({ foo: 'bar ' }))
//...

const appB = new Hono()
  .get('/1', (c) => c.json({ foo: 'bar ' }))
  .get('/2', (c) => c.json({ foo: 'bar ' }))
//...

const clientA = hc<typeof appA>('http://localhost/a')
const clientB = hc<typeof appB>('http://localhost/b')
```

**Reply by bcheung** [reply]
btw you can still join the actual hono applications into one big app and deploy one app. the part that fails is the joining of the apps and inferring the routes into the single client.

You just need to split up the clients, not the actual hono app. The actual hono app can still be merged together because the failure only occurs when you create the client and add the route types.

I have a clearer example of how I did this here:
https://github.com/honojs/hono/issues/2399#issuecomment-2675421823

**Comment by yusukebe** [comment, marked as answer]
Hi @RomanNabukhotnyi 

I've added fix #2412, and the performance has improved a little. Then, it does not throw the error in your app:

<img width="676" alt="Screenshot 2024-03-25 at 18 56 08" src="https://github.com/honojs/hono/assets/10682/e9f7750b-8b57-4e05-8c86-193b901d4688">

But there is still improvement to be made as the `hc` function is not yet smart enough!

**Reply by RomanNabukhotnyi** [reply]
Thanks!

**Comment by shutt** [comment]
In my case was a problem with JSONB type, check if you have any and try to change it.
